docs/index-type-operator.md

```diff
@@ -0,0 +1,88 @@
+# `index` type operators
+
+## Summary
+
+This RFC proposes the addition of one type operator, `index`, which can be used to look up a specific property of another type (like TypeScript's Indexed Access Type).
+
+## Motivation
+
+The primary motivation of this proposal is to allow Luau to define a type by accessing a property of another existing type. For instance, consider the following example code:
+```lua
+type Person = {
+  age: number,
+  name: string,
+  alive: boolean
+}
+
+local bob: Person = {
+  age = 22,
+  name = "Bob",
+  alive = true
+}
+
+local function doSmt(param: typeof(bob["age"])) -- param = string
+  -- rest of code
+end
+
+type unionType = typeof(bob["age"]) | typeof(bob["name"]) | typeof(bob["alive"]) -- unionType = number | string | boolean
+```
+
+This is a valid Luau program; however, in order to define the type of `Person["age"]` we had to first declare a variable `bob` and utilize the `typeof` type operator. This is quite cumbersome when developers want to typecheck using the type of `Person["age"]` without having to declare a variable first. Additionally, in order to define the union type of all the properties of `Person`, current Luau requires an explicit list of each property using `typeof`.
+
+The expected outcome of the index type operator is that it will enhance the developing experience and allow the Roblox community to develop better type-checked programs.
+
+## Design
+
+The solution to this problem is a type operator, `index`, that can compute the type based on the static properties of `Person`. Formally, the `index` type operator will take in two arguments: the type to index and the type to index it with. This would allow us to instead write the following code:
+```lua
+type Person = {
+  age: number,
+  name: string,
+  alive: boolean
+}
+
+local function doSmt(param: index<Person, "age">) -- param = string
+  -- rest of code
+end
+
+type idxType = index<Person, keyof<Person>> -- idxType = number | string | boolean
+
+type idxType2 = index<Person, "age" | "name"> -- idxType2 = number | string
+```
+
+Now, the type of `doSmt()`'s parameter can be defined without declaring a variable `bob`. Additionally, regardless of how the type `Person` grows, `idxType` will always be defined as the union of all the properties.
+
+Error messages will be displayed for incorrect syntax. If the given type used to access a property is invalid, 
+```lua
+type age = index<Person, "ager"> -- Error message: Property 'ager' does not exist on type 'Person'.
+```
+If the given value used to access a property is not a type,
+```lua
+local key = "age"
+type age = index<Person, key> -- Error message: Type 'key' cannot be used as an index type.
+```
```

Comment:
It's worth probably stating explicitly since this type operator will be built into the type family system, these errors will be part of the existing errors for type family reduction errors. This sort of specific message is fine though, as the stringification for type family reduction errors already specializes the errors for the existing type operators for the sake of better DX.